Support for authenticating using Service Principal with a Client Certificate (#2471)
terraform-providers/terraform-provider-azurerm: website/docs/auth/service_principal_client_secret.html.markdown

---
layout: "azurerm"
page_title: "Azure Provider: Authenticating via a Service Principal and a Client Secret"
sidebar_current: "docs-azurerm-authentication-service-principal-client-secret"
description: |-
  This guide will cover how to use a Service Principal (Shared Account) with a Client Secret as authentication for the Azure Provider.

---

# Azure Provider: Authenticating using a Service Principal with a Client Secret

Terraform supports a number of different methods for authenticating to Azure:

* [Authenticating to Azure using the Azure CLI](azure_cli.html)
* [Authenticating to Azure using Managed Service Identity](managed_service_identity.html)
* [Authenticating to Azure using a Service Principal and a Client Certificate](service_principal_client_certificate.html)
* Authenticating to Azure using a Service Principal and a Client Secret (which is covered in this guide)

---

> We recommend using either a Service Principal or Managed Service Identity when running in automation (for example on a CI Server) - and authenticating using the Azure CLI when running Terraform locally.

Reviewer: This wording confuses me "...running in automation...", maybe "...running Terraform via automation..." or something along those lines?

Reviewer 2: perhaps instead of automation we could use the term non-interactive?

$ az account list-locations
```

> Finally, since we're logged into the Azure CLI as a Service Principal we recommend logging out of the Azure CLI (but you can instead log in using your user account):

Reviewer: If we are just going to tell the user to log out as the Service Principal, why did we tell them to log in as the Service Principal in the first place? I find this part a little confusing as well "(but you can instead log in using your user account)"...

Author: this exists to validate the credentials being used have permissions to the subscription before moving onto the next step (configuring those credentials in Terraform). We used to have a lot of issues opened where the Service Principal didn't have permissions to the subscription, adding this step in mostly resolved these tbh

```bash
$ az logout
```

Information on how to configure the Provider block using the newly created Service Principal credentials can be found below.

---

### Creating a Service Principal in the Azure Portal

> There are three tasks needed to create a Service Principal via [the Azure Portal](https://portal.azure.com):

Reviewer: ```suggestion
There are three tasks that need to be completed in order to create a Service Principal via [the Azure Portal](https://portal.azure.com):
```

Reviewer 2: ```suggestion
There are three tasks required to create a Service Principal via [the Azure Portal](https://portal.azure.com):
```
?

Author: closing in favour of the suggestion below

> There are three tasks needed to create a Service Principal via [the Azure Portal](https://portal.azure.com):

Author: on reflection I think this'd read better as - WDYT:

```suggestion
There are three tasks necessary to create a Service Principal using [the Azure Portal](https://portal.azure.com):
```

Reviewer 2: think thats best

 1. Create an Application in Azure Active Directory (which acts as a Service Principal)
 2. Generating a Client Secret for the Azure Active Directory Application (which can be used for authentication)
 3. Grant the Application access to manage resources in your Azure Subscription

### 1. Creating an Application in Azure Active Directory

> Firstly navigate to [the **Azure Active Directory** overview](https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/Overview) within the Azure Portal - [then select the **App Registration** blade](https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps/RegisteredApps/Overview) and click **Endpoints** at the top of the **App Registration** blade. A list of URIs will be displayed and you need to locate the URI for **OAUTH 2.0 AUTHORIZATION ENDPOINT** which contains a GUID. This is your Tenant ID / the `tenant_id` field mentioned above.

Reviewer: Does "This is your Tenant ID (the tenant_id field mentioned above)." seem clearer than "This is your Tenant ID / the tenant_id field mentioned above."?

Author: (as above)

- **Name** - this is a friendly identifier and can be anything (e.g. "Terraform")
- **Application Type** - this should be set to "Web app / API"
- **Sign-on URL** - this can be anything, providing it's a valid URI (e.g. https://terra.form)

At this point the newly created Azure Active Directory application should be visible on-screen - if it's not, navigate to the [the **App Registration** blade](https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps/RegisteredApps/Overview) and select the Azure Active Directory application. At the top of this page, the "Application ID" GUID is the `client_id` you'll need.

### 2. Generating a Client Secret for the Azure Active Directory Application

Now that the Azure Active Directory Application exists we can create a Client Secret which can be used for authentication - to do this select **Settings** and then **Keys**. This screen displays the Passwords (Client Secrets) and Public Keys (Client Certificates) which are associated with this Azure Active Directory Application.

> On this screen we can generate a new Password by entering a Description and selecting an Expiry Date, and then pressing **Save**. Once the Password has been generated it'll be displayed on screen - _however it's only displayed one time_ so **be sure to copy it now** (otherwise you'll need to regenerate a new key). This newly generated Password forms the `client_secret` you'll need.

Reviewer: ```suggestion
On this screen we can generate a new Password by entering a Description and selecting an Expiry Date, and then pressing **Save**. Once the Password has been generated it will be displayed on screen - _the Password is only displayed once_ so **be sure to copy it now** (otherwise you will need to regenerate a new key). This newly generated Password is the `client_secret` you will need.
```
